### PR TITLE
Nystrom

### PR DESCRIPTION
**Pull request by DerrickhSun**
Implemented the Nystrom approximation method. It uses the Moore-Penrose pseudoinverse, which the original paper gives an iterative approach to calculating. The iterative approach is still a work-in-progress, but we can try both their approach and the pytorch pseudo-inverse and add the comparison to the paper once it's done.

### CODE REVIEW

File: src/conf/linear_regression_nystrom.yaml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+training:
+    task: linear_regression
+    curriculum:
+        points:
+            start: 11
+            end: 41
+            inc: 2
+            interval: 2000
+
+out_dir: ../models/linear_regression_relu_attn
```
Comment (nelson-lojo) [inline src/conf/linear_regression_nystrom.yaml:16]: This should be `../models/linear_regression_nystrom`

File: src/models.py [hunk 2]
```diff
@@ -185,6 +253,20 @@ def __init__(self, *args, **kwargs):
                 attn_module_class
             )
 
+class TransformerNystrom(TransformerModel):
+    def __init__(self, *args, **kwargs):
+        super(TransformerNystrom, self).__init__(*args, **kwargs)
+
+        # Override the attention mechanism with one that replaces softmax with ReLU
```
Comment (nelson-lojo) [inline src/models.py:260]: minor change: this class isn't relu

File: src/models.py [hunk 3]
```diff
@@ -171,6 +182,63 @@ def relu_attn(self, query, key, value, attention_mask=None, head_mask=None):
...
+
+    # this was taken directly from the nystrom paper
+    einops_eq = '... i d, ... j d -> ... i j'
+    sim1 = einsum(einops_eq, query, k_bar)
+    sim2 = einsum(einops_eq, q_bar, k_bar)
+    sim3 = einsum(einops_eq, q_bar, key)
+
+    # this was also taken from the nystrom paper, but modified
+    attn1, attn2, attn3 = map(lambda t: t.softmax(dim = -1), (sim1, sim2, sim3))
+    attn2_inv = torch.linalg.pinv(attn2)
```
Comment (nelson-lojo) [inline src/models.py:231]: Reply to this message with the speed measurements -- basically what's the speedup of `torch.linalg.pinv` over the nystrom `pinv` for varying values of `m`? (do just like 2-3 runs for m=2,5,8)

File: src/models.py [hunk 4]
```diff
@@ -7,6 +7,8 @@
 import warnings
 from sklearn import tree
 import xgboost as xgb
+import math
+from torch import nn, einsum
```
Comment (nelson-lojo) [inline src/models.py:11]: `nn` already imported

**Review by nelson-lojo — CHANGES_REQUESTED** [review]
minor changes; documentation

File: src/conf/linear_regression_nystrom.yaml (hunk 1 above)
Comment (nelson-lojo) [reply src/conf/linear_regression_nystrom.yaml:16]: no wait it should be `../models/linear_regression`

File: src/models.py (hunk 3 above)
Comment (DerrickhSun) [reply src/models.py:231]: ### training time using 51 steps:
**pytorch method:**
m=8: 1:13
m=5: 1:11
m=2: 1:07
**iterative method with 7 iterations:**
m=8: 1:21
m=5: 1:15
m=2: 1:08

**Review by nelson-lojo — COMMENTED** [review]
looks good! I'll run a few gradient steps to make sure
